global-hooks/migrate/migrate_modules.go

```diff
@@ -0,0 +1,150 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hooks
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	"github.com/flant/addon-operator/sdk"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/deckhouse/deckhouse/go_lib/dependency"
+	"github.com/deckhouse/deckhouse/go_lib/hooks/ensure_crds"
+)
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	OnStartup: &go_hook.OrderedConfig{Order: 10},
+}, dependency.WithExternalDependencies(modulesCRMigrate))
+
+func modulesCRMigrate(input *go_hook.HookInput, dc dependency.Container) error {
+	kubeCl, err := dc.GetK8sClient()
+	if err != nil {
+		return fmt.Errorf("cannot init Kubernetes client: %v", err)
+	}
+
+	modulesMigrationCM := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "modules-cr-names-migration",
+			Namespace: "d8-system",
+		},
+	}
+
+	_, err = kubeCl.CoreV1().ConfigMaps(modulesMigrationCM.Namespace).Get(context.TODO(), modulesMigrationCM.Name, metav1.GetOptions{})
+	if err == nil {
+		input.LogEntry.Info("Modules migration configmap exists, skipping the migration")
+		return nil
+	}
+	if !errors.IsNotFound(err) {
+		return err
+	}
+
+	// old
+	emsGVR := schema.ParseGroupResource("externalmodulesources.deckhouse.io").WithVersion("v1alpha1")
+	emrGVR := schema.ParseGroupResource("externalmodulereleases.deckhouse.io").WithVersion("v1alpha1")
+	// new
+	msGVR := schema.ParseGroupResource("modulesources.deckhouse.io").WithVersion("v1alpha1")
+	mrGVR := schema.ParseGroupResource("modulereleases.deckhouse.io").WithVersion("v1alpha1")
+
+	skipMigrationCount := 0
+
+	moduleSources, err := kubeCl.Dynamic().Resource(emsGVR).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+
+		input.LogEntry.Info("ExternalModuleSource resource is not in the cluster")
+		skipMigrationCount++
+	}
+
+	moduleReleases, err := kubeCl.Dynamic().Resource(emrGVR).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+
+		input.LogEntry.Info("ExternalModuleRelease resource is not in the cluster")
+		skipMigrationCount++
+	}
+
+	if skipMigrationCount == 2 { // both resources are absent
+		input.LogEntry.Info("Skipping modules migration")
+		return nil
+	}
+
+	ensureRes := ensure_crds.EnsureCRDs("/deckhouse/modules/005-external-module-manager/crds/module-*.yaml", input, dc)
```

Comment:
Couldn't we move this hook to the 005-external-module-manager module. Because installing module crds in the global hook looks really weird and breaks all kind of encapsulation.

Reply:
Good catch; it is not apparent. I added a comment to the migration hook.